Bridge/src/zeta.cpp

```diff
@@ -0,0 +1,120 @@
+#include "zeta.hpp"
+
+#include <chrono>
+#include <iostream>
+
+#include "debug.hpp"
+
+ZetaRfRadio::ZetaRfRadio()
+{
+    std::cout << "Starting Zeta Radio..." << std::endl;
+
+    if (!this->zeta_.beginWithPacketLengthOf(PACKET_LENGTH))
+        throw "ZetaRf begin failed. Check wiring?";
+
+    if (!this->zeta_.startListeningSinglePacketOnChannel(ZetaRfRadio::ZETA_CHANNEL))
+        throw "ZetaRf startListening failed.";
+
+    debug << "Starting ZetaRf worker thread..." << std::endl;
+    this->worker_ = std::thread(&ZetaRfRadio::rx_tx_loop, this);
+
+    std::cout << "ZetaRf init done." << std::endl;
+}
+
+ZetaRfRadio::~ZetaRfRadio()
+{
+    if (this->worker_.joinable())
+    {
+        this->should_worker_join_ = true;
+        this->worker_.join();
+    }
+}
+
+void ZetaRfRadio::set_on_received(std::function<void(Frame)> callback)
+{
+    this->on_receive_ = callback;
+}
+
+void ZetaRfRadio::send_packets(const std::vector<Frame> frames)
+{
+    this->send_queue_.push(frames);
+}
+
+void ZetaRfRadio::rx_tx_loop()
+{
+    while (!this->should_worker_join_)
+    {
+        // Receive packet
+        this->process_zeta_events();
+        if (this->zeta_.hasDataAvailable())
+            this->read_packet();
+
+        // Transmit next queued packet.
+        // We could send all queue packets, but we don't want to overwhelm the
+        // receiver's rx fifo if it is also transmitting.
+        if (auto packet = this->send_queue_.pop())
+        {
+            debug << "Sending packet: " << &*packet << std::endl;
+            this->transmit_packet(*packet);
+        }
+
+        std::this_thread::sleep_for(std::chrono::milliseconds(2));
+    }
+    debug << "ZetaRf worker thread exiting..." << std::endl;
+}
+
+void ZetaRfRadio::process_zeta_events()
+{
+    const auto event = this->zeta_.checkForEvent();
+
+    if (!event)
+        return;
+
+    if (event & ZetaRf::Event::DeviceBusy)
```

Comment:
We can simply just do: `if (ZetaRf::Event::DeviceBusy)` here now because we check if `event` is falsy above!

Reply:
I think you've missed what's happening here :wink: The events here use bit flags!

`ZetaRf::Event` is an enum type, of which `DeviceBusy` is a possible value. `DeviceBusy`, and the rest of the enums, all have a (compile-time constant) integer value, which are all powers of two (see [here](https://github.com/GIPdA/ZetaRF/blob/v2/src/zetarf_radio.hpp) again). So, all the values will, in binary, be all `0` except for a single `1`.

`event` is also an integer. If we look at its binary representation, it is able to represent any (and every) combination of ZetaRf events by simply summing all the event values we wish for it to contain. Lets say `event` was to contain `ZetaRf::Event::PacketReceived`, which has a value of `0b000 0100`. This would mean that `event` could be any integer, as long as the third last bit is a `1`, to indicate that `PacketReceived` event.

So, on these lines when we're checking for events, we actually take a bitewise and of `event` and the binary number which represents the event we're checking. So, `event & ZetaRf::Event::PacketReceived` will either be `0b0000 0100` if there's been a `PacketReceived` event, or `0b0000 0000` otherwise. The former is truthy and the latter is falsey, so we can use this as the condition for an if statement.

In summary, `event & ZetaRf::Event::PacketReceived` is how we actually check if an event has occurred, `ZetaRf::Event::DeviceBusy` on its own is just a constant enum/number. I realise now you probably know how this all works and just missed that it's a bitewise `&` and didn't know `Event` was an enum... I hope someone else finds this interesting though! It's called bit flags if you want to google.

Reply:
Thanks for writing that all out! Definitely missed that it was a bitwise `and` and I was still stuck in the javascript heavens. But thanks for clearing that up, I always find bit operators quite interesting and a cool way of doing things.

Reply:
Yeah they're super cool! Stores 32 booleans in the space of one integer!